### PR TITLE
rustdoc: refactor: centralize all command-line argument parsing

### PR DESCRIPTION
**Pull request by QuietMisdreavus**
This is something i've wanted to do for a while, since we keep having to add new arguments to places like `rust_input` or `core::run_core` whenever we add a new CLI flag or the like. Those functions have inflated up to 11-19, and in some cases hiding away the locations where some CLI flags were being parsed, obscuring their use. Now, we have a central place where all command-line configuration occurs, including argument validation.

One note about the design: i grouped together all the arguments that `html::render::run` needed, so that i could pass them on from compilation in one lump instead of trying to thread through individual items or clone the entire blob ahead of time.

One other thing this adds is that rustdoc also now recognizes all the `-Z` options that rustc does, since we were manually grabbing a few previously. Now we parse a full `DebuggingOptions` struct and hand it directly to rustc when scraping docs.

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
That simplifies the code a lot!

r=me once CI passed.

**Comment by QuietMisdreavus** [comment]
Well that was an intense rebase. @GuillaumeGomez can you take another look? I had to change around `html::render::{run, write_shared}` a little more, to make sure that `markdown::render` could be called properly. I moved `(enable_)index_page` out of the Context, since i had to put them in `RenderOptions` anyway, and i changed `markdown::render` to take the input file separately so it could take a `RenderOptions` instead. To make that work, i moved the markdown-specific items into `RenderOptions`, too. I ran tests locally and everything seems to still work.

File: src/librustdoc/config.rs [hunk 1]
```diff
@@ -0,0 +1,520 @@
...
+
+    let removed_flags = [
+        "plugins",
+        "plugin-path",
+    ];
+
+    for &flag in removed_flags.iter() {
+        if matches.opt_present(flag) {
+            diag.struct_warn(&format!("the '{}' flag no longer functions", flag))
+                .warn("see CVE-2018-1000622")
```
Comment (GuillaumeGomez) [inline src/librustdoc/config.rs]: All removed flags are linked to this CVE?

File: src/librustdoc/config.rs [hunk 2]
```diff
@@ -0,0 +1,520 @@
...
+                .emit();
+        }
+    }
+}
+
+/// Extracts `--extern-html-root-url` arguments from `matches` and returns a map of crate names to
+/// the given URLs. If an `--extern-html-root-url` argument was ill-formed, returns an error
+/// describing the issue.
+fn parse_extern_html_roots(matches: &getopts::Matches)
+    -> Result<BTreeMap<String, String>, &'static str>
```
Comment (GuillaumeGomez) [inline src/librustdoc/config.rs]: If you're using this syntax, at least use it on the whole function header ;) :

```
fn parse_extern_html_roots(
    matches: &getopts::Matches,
) -> Result<BTreeMap<String, String>, &'static str>
```

File: src/librustdoc/config.rs [hunk 3]
```diff
@@ -0,0 +1,520 @@
...
+    // Options that alter generated documentation pages
+
+    /// Crate version to note on the sidebar of generated docs.
+    pub crate_version: Option<String>,
+    /// Collected options specific to outputting final pages.
+    pub render_options: RenderOptions,
+}
+
+/// Configuration options for the HTML page-creation process.
+#[derive(Clone)]
```
Comment (GuillaumeGomez) [inline src/librustdoc/config.rs]: Please also add `Debug`. I have the feeling that I'll need it.

File: src/librustdoc/config.rs [hunk 4]
```diff
@@ -0,0 +1,520 @@
...
+use core::new_handler;
+use externalfiles::ExternalHtml;
+use html;
+use html::markdown::IdMap;
+use opts;
+use passes::{self, DefaultPassOption};
+use theme;
+
+/// Configuration options for rustdoc.
+#[derive(Clone)]
```
Comment (GuillaumeGomez) [inline src/librustdoc/config.rs]: Please also add `Debug`.

Comment (QuietMisdreavus) [reply src/librustdoc/config.rs]: I can't derive Debug on the global Options Struct without also adding it to rustc's CodegenOptions and DebuggingOptions. (And when i tried adding it to those structs just now, i would need to add it to other types for that to work out.) Do you want me to make a manual Debug impl that skips those fields?

File: src/librustdoc/config.rs (hunk 1 above)
Comment (QuietMisdreavus) [reply src/librustdoc/config.rs]: Both of these flags are, yes. They showed a similar message that was printed with `eprintln!()`, which was being checked in `rust_input`.

File: src/librustdoc/config.rs (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustdoc/config.rs]: Yes please!

**Comment by QuietMisdreavus** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
Thanks!

@bors: r+

**Pull request merged**